core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+   * It either invokes the desired component action or completes the composition invocation.
+   * It also checks the invocation counts against the limits.
+   *
+   * @param user the identity invoking the action
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeConductor(user: Identity, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) = {
+
+    if (session.accounting.conductors > 2 * actionSequenceLimit) {
+      // composition is too long
+      val response = ActivationResponse.applicationError(compositionIsTooLong)
+      completeAppActivation(user, session, response)
+    } else {
+      // inject state into payload if any
+      val params = session.state
+        .map { state =>
+          Some(JsObject(payload.getOrElse(JsObject()).fields ++ state.fields))
+        }
+        .getOrElse(payload)
+
+      // invoke conductor action
+      session.accounting.conductors += 1
+      val response =
+        invokeSimpleAction(
+          user,
+          session.action,
+          params,
+          Some(session.action.limits.timeout.duration + 1.minute), // wait for result
+          Some(session.activationId)) // cause is session id
+
+      response.onComplete {
+        case Failure(t) =>
+          // invocation failure
+          val response = ActivationResponse.whiskError(compositionActivationFailure)
+          completeAppActivation(user, session, response)
+        case Success(Left(activationId)) =>
+          // invocation timeout
+          session.logs += activationId
+          val response = ActivationResponse.whiskError(compositionActivationTimeout(activationId))
+          completeAppActivation(user, session, response)
+        case Success(Right(activation)) =>
+          // successful invocation
+          session.logs += activation.activationId
+          session.duration += activation.duration.getOrElse(activation.end.toEpochMilli - activation.start.toEpochMilli)
+
+          val result = activation.resultAsJson
+
+          // extract params from result
+          val params = result.getFields(WhiskActivation.paramsField).headOption.map { p =>
+            Try(p.asJsObject).getOrElse(JsObject(WhiskActivation.valueField -> p)) // ensure params is a dictionary
+          }
+
+          // update session state
+          session.state = result.getFields(WhiskActivation.stateField).headOption.flatMap { p =>
+            Try(Some(p.asJsObject)).getOrElse(None)
+          }
+
+          // extract next action from result and invoke
+          result.getFields(WhiskActivation.actionField).headOption match {
+            case None =>
+              // no next action, end composition execution, return to caller
+              val response = ActivationResponse(activation.response.statusCode, Some(params.getOrElse(result)))
```

Comment:
why is the response result `params` else `result` and not just result?

Reply:
I don't want to restrict the output schema of component actions. For instance, a component action should be permitted to return `{ action: 'foo' }` or `{ state: 'OK' }`. Hence, a composition should be permitted to return `{ action: 'foo' }` without triggering the execution of action `foo`. This is made possible by the code above (`params` case). A conductor may return `{ action: 'foo' }` to trigger the execution of action `foo` or `{ params: { action: 'foo' } }` to return `{ action: 'foo' }` without further processing. The default case (`result` case) is intended to handle error responses from conductor actions.